### ISSUE TITLE
card-dnie: OpenSC FTBFS on (k)freebsd and hurd

### ISSUE THREAD

**Issue by bigon** [OpenSC/OpenSC#1366]
Hi,

It seems that OpenSC FTBFS on (k)freebsd and hurd due to an explicit `#error` in the dnie driver code

Debian kfreebsd and hurd contains the `pinentry` executable, so I don't see the reason why the compilation should fail.

**Comment by bigon** [comment]
Well looking more at the code, the `#ifdef linux` seems to cover everything that is not Apple or Win32.

IMVHO, these tests should be changed to something more specific, like is it using the glibc or something. That would of course not cover the real freebsd

**Comment by dengert** [comment]
It looks like card-dnie.c is trying to handle prompting for a PIN when it finds a key with `user_consent`.

At least with PKCS\#11, other cards hande `user_consent` keys by setting the `CKA_ALWAYS_AUTHENTICATE`  so the calling application will spot this and prompt for PIN, then do `C_Login(CKU_CONTEXT_SPECIFIC)` which turns into a verify command. 

Thus the driver does not need to do any prompting. 

Is it time use the above in card-dnie.c in place of internal prompting?

**Comment by bigon** [comment]
You mean treating the `linux` case as the default/else case? or Checking if we are building against glibc?

**Comment by frankmorgner** [comment]
yes. But I think glibc isn't even needed. As far as I can see, the communication to the external program for user consent is standard-C.

**Comment by frankmorgner** [comment]
By the way, is there an example program for `user_consent`?

**Comment by dengert** [comment]
The need for a OK appears to be a policy but the card has no way to enforce such a policy. 
PKCS\#11 says a CKU_CONTEXT_SPECIFIC login  (when used in the context of a  Sign operation) a   
Re-authentication. But the card-dnie.c does not re-authenticate, only requires some positive user response.  So for the card-dnie.c that  could a  CKU_CONTEXT_SPECIFIC login with a PIN value of "OK", "1" or the actual PIN.  
(Security for the card could actually be improved, if the actual pin was required, and the driver did a verify to prove it is the actual PIN.

I only question the need of  pages of code that are OS specific in a card driver. 
Are there any other card drives that do something similar? 
If so should they  use a common set of OS specific code? 
If not, can CKU_CONTEXT_SPECIFIC login  by used instead?
